nntrainer/layers/preprocess_flip_layer.cpp

```diff
@@ -26,55 +27,80 @@ const std::string PreprocessFlipLayer::flip_vertical = "vertical";
-  output_dim = input_dim;
+void PreprocessFlipLayer::finalize(InitLayerContext &context) {
+  context.setOutputDimensions(context.getInputDimensions());
 
   rng.seed(getSeed());
   flip_dist = std::uniform_real_distribution<float>(0.0, 1.0);
+}
+
+void PreprocessFlipLayer::setProperty(const std::vector<std::string> &values) {
+  /// @todo: deprecate this in favor of loadProperties
+  for (unsigned int i = 0; i < values.size(); ++i) {
+    std::string key;
+    std::string value;
+    std::stringstream ss;
+
+    if (getKeyValue(values[i], key, value) != ML_ERROR_NONE) {
+      throw std::invalid_argument("Error parsing the property: " + values[i]);
+    }
+
+    if (value.empty()) {
+      ss << "value is empty: key: " << key << ", value: " << value;
+      throw std::invalid_argument(ss.str());
+    }
 
-  return ML_ERROR_NONE;
+    /// @note this calls derived setProperty if available
+    setProperty(key, value);
+  }
 }
 
-void PreprocessFlipLayer::setProperty(const PropertyType type,
+void PreprocessFlipLayer::setProperty(const std::string &type_str,
                                       const std::string &value) {
+  using PropertyType = LayerV1::PropertyType;
+  LayerV1::PropertyType type =
+    static_cast<LayerV1::PropertyType>(parseLayerProperty(type_str));
+
   switch (type) {
-  case PropertyType::flip_direction:
-    if (!value.empty()) {
-      if (istrequal(value, flip_horizontal)) {
-        flipdirection = FlipDirection::horizontal;
-      } else if (istrequal(value, flip_vertical)) {
-        flipdirection = FlipDirection::vertical;
-      } else if (istrequal(value, flip_horizontal_vertical)) {
-        flipdirection = FlipDirection::horizontal_and_vertical;
-      } else {
-        throw std::invalid_argument("Argument flip direction is invalid");
-      }
+  case PropertyType::flip_direction: {
+    if (istrequal(value, flip_horizontal)) {
+      flipdirection = FlipDirection::horizontal;
+    } else if (istrequal(value, flip_vertical)) {
+      flipdirection = FlipDirection::vertical;
+    } else if (istrequal(value, flip_horizontal_vertical)) {
+      flipdirection = FlipDirection::horizontal_and_vertical;
+    } else {
+      throw std::invalid_argument("Argument flip direction is invalid");
     }
-    break;
+  } break;
   default:
-    LayerV1::setProperty(type, value);
-    break;
+    std::string msg =
+      "[PreprocessFlipLayer] Unknown Layer Property Key for value " +
+      std::string(value);
+    throw exception::not_supported(msg);
   }
 }
 
-void PreprocessFlipLayer::forwarding(bool training) {
+void PreprocessFlipLayer::forwarding(RunLayerContext &context, bool training) {
   if (!training) {
-    for (unsigned int idx = 0; idx < input_dim.size(); idx++) {
-      net_hidden[idx]->getVariableRef() = net_input[idx]->getVariableRef();
+    for (unsigned int idx = 0; idx < context.getNumInputs(); idx++) {
+      context.getOutput(idx) = context.getInput(idx);
```

Comment:
not related to the PR but is this allowed in the point of graph?

Reply:
Ideally this should do copy or set supportInPlace() to true.
I will add a todo to check this later.

Reply:
This just does not use the given memory and works differently for train/inference.
Maybe `supportInPlace` should take `ExecutionMode` as an argument.

Reply:
I'll keep in mind :)